### ISSUE TITLE
CPD Exception thrown: java.lang.OutOfMemoryError

### ISSUE THREAD

**Issue by mathieugouin** [pmd/pmd#1417]
**Affects PMD Version:** 6.8.0

**Rule:**

```
$ run.sh cpd --minimum-tokens 150 --language cpp --format xml --failOnViolation false --files ProjectRoot --exclude ProjectRoot/Unwanted $( find -type d \( -name "Test" -o -name "VNV" \) | sed 's/^\.\//^C
```

**Description:**

A java exception is raised:

```
Exception in thread "main" java.lang.OutOfMemoryError: Java heap space
        at java.util.Arrays.copyOf(Unknown Source)
        at java.lang.AbstractStringBuilder.expandCapacity(Unknown Source)
        at java.lang.AbstractStringBuilder.ensureCapacityInternal(Unknown Source)
        at java.lang.AbstractStringBuilder.append(Unknown Source)
        at java.lang.StringBuilder.append(Unknown Source)
        at java.io.WinNTFileSystem.slashify(Unknown Source)
        at java.io.WinNTFileSystem.resolve(Unknown Source)
        at java.io.File.getAbsolutePath(Unknown Source)
        at net.sourceforge.pmd.cpd.SourceCode$FileCodeLoader.getFileName(SourceCode.java:120)
        at net.sourceforge.pmd.cpd.SourceCode.getFileName(SourceCode.java:211)
        at net.sourceforge.pmd.cpd.CPPTokenizer.tokenize(CPPTokenizer.java:71)
        at net.sourceforge.pmd.cpd.CPD.addAndThrowLexicalError(CPD.java:142)
        at net.sourceforge.pmd.cpd.CPD.add(CPD.java:137)
        at net.sourceforge.pmd.cpd.CPD.add(CPD.java:107)
        at net.sourceforge.pmd.cpd.CPD.add(CPD.java:69)
        at net.sourceforge.pmd.cpd.CPD.addDirectory(CPD.java:79)
        at net.sourceforge.pmd.cpd.CPD.addRecursively(CPD.java:64)
        at net.sourceforge.pmd.cpd.CPDCommandLineInterface.addSourcesFilesToCPD(CPDCommandLineInterface.java:133)
        at net.sourceforge.pmd.cpd.CPDCommandLineInterface.addSourceFilesToCPD(CPDCommandLineInterface.java:113)
        at net.sourceforge.pmd.cpd.CPDCommandLineInterface.main(CPDCommandLineInterface.java:86)
        at net.sourceforge.pmd.cpd.CPD.main(CPD.java:177)

```

**Code Sample demonstrating the issue:**

```

```

**Running PMD through:** *CLI + cygwin*

OS: Windows 7 64 Bit

$ java -version
java version "1.8.0_73"
Java(TM) SE Runtime Environment (build 1.8.0_73-b02)
Java HotSpot(TM) Client VM (build 25.73-b02, mixed mode)

**Comment by jsotuyod** [comment]
@mathieugouin thanks for the report.

This is an interesting case. CPD is a memory-hungry process, but at the point you are seeing this, it hasn't even started analysis...

I've several questions to dig deeper into this:
1. how many files are in your project?
2. have you tried overriding the default heapsize? Try `$ HEAPSIZE=4G run.sh cpd …` (4 or whatever value you see fit)
3. could you run `java -XX:+PrintFlagsFinal -version | grep HeapSize` and share the results?

Thanks in advance

**Comment by mathieugouin** [comment]
Hello @jsotuyod 

Thanks for the response.  I found out that we had a huge file containing an image raw structure in c.  This was for debug purposes and was not excluded from cpd.  This is was caused the crash.  We excluded the file, and now it works like a charm.

You can close this issue if you want, or let me know if you are still interested in having more information as requested.

**Comment by jsotuyod** [comment]
@mathieugouin oh, great! I'll close this for the time being then. We will probably revisit CPD at some point.
